p2p/transport_mconn.go

```diff
@@ -0,0 +1,646 @@
+	// by the router once we rewrite the P2P core.
+	conns       ConnSet
+	connFilters []ConnFilterFunc
+}
+
+// NewMConnTransport sets up a new MConn transport.
+func NewMConnTransport(
+	logger log.Logger,
+	nodeInfo NodeInfo, // FIXME should use DefaultNodeInfo, left for code compatibility
+	privKey crypto.PrivKey,
+	mConnConfig conn.MConnConfig,
+	opts ...MConnTransportOption,
+) *MConnTransport {
+	m := &MConnTransport{
+		privKey:      privKey,
+		nodeInfo:     nodeInfo.(DefaultNodeInfo),
+		mConnConfig:  mConnConfig,
+		channelDescs: []*ChannelDescriptor{},
+
+		dialTimeout:      defaultDialTimeout,
+		handshakeTimeout: defaultHandshakeTimeout,
+		filterTimeout:    defaultFilterTimeout,
+
+		logger:   logger,
+		chAccept: make(chan *mConnConnection),
+		chError:  make(chan error),
+		chClose:  make(chan struct{}),
+
+		conns:       NewConnSet(),
+		connFilters: []ConnFilterFunc{},
+	}
+	for _, opt := range opts {
+		opt(m)
+	}
+	return m
+}
+
+// SetChannelDescriptors implements Transport.
+func (m *MConnTransport) SetChannelDescriptors(chDescs []*conn.ChannelDescriptor) {
+	m.channelDescs = chDescs
+}
+
+// Listen asynchronously listens for inbound connections on the given endpoint.
+// It must be called exactly once before calling Accept(), and the caller must
+// call Close() to shut down the listener.
+func (m *MConnTransport) Listen(endpoint Endpoint) error {
+	if m.listener != nil {
+		return errors.New("MConn transport is already listening")
+	}
+	err := m.normalizeEndpoint(&endpoint)
+	if err != nil {
+		return fmt.Errorf("invalid MConn listen endpoint %q: %w", endpoint, err)
+	}
+
+	m.listener, err = net.Listen("tcp", fmt.Sprintf("%v:%v", endpoint.IP, endpoint.Port))
+	if err != nil {
+		return err
+	}
+	if m.maxIncomingConnections > 0 {
+		m.listener = netutil.LimitListener(m.listener, m.maxIncomingConnections)
+	}
+
+	// Spawn a goroutine to accept inbound connections asynchronously.
+	go m.accept()
+
+	return nil
+}
+
+// accept accepts inbound connections in a loop, and asynchronously handshakes
+// with the peer to avoid head-of-line blocking. Established connections are
+// passed to Accept() via the channel m.chAccept.
+// See: https://github.com/tendermint/tendermint/issues/204
+func (m *MConnTransport) accept() {
+	for {
+		tcpConn, err := m.listener.Accept()
+		if err != nil {
+			select {
+			case m.chError <- err:
+			case <-m.chClose:
+			}
+			return
+		}
+		go func() {
+			err := m.filterTCPConn(tcpConn)
+			if err != nil {
+				_ = tcpConn.Close()
+				select {
+				case m.chError <- err:
+				case <-m.chClose:
+				}
+			}
+			conn, err := newMConnConnection(m, tcpConn, "")
+			if err != nil {
+				m.conns.Remove(tcpConn)
+				_ = tcpConn.Close()
+				select {
+				case m.chError <- err:
+				case <-m.chClose:
+				}
+			} else {
+				select {
+				case m.chAccept <- conn:
+				case <-m.chClose:
+					_ = conn.Close()
+				}
+			}
+		}()
+	}
+}
+
+// Accept implements Transport.
```

Comment:
Can we elaborate a bit on this godoc? e.g. When and who calls `Accept` and what are the preliminary requirements to get a `Connection` returned (i.e. what triggers `accept` to send a conn)?

Reply:
Interface methods should be documented on the interface itself, i.e. `Transport` -- there's no point duplicating that information across all of the implementations, and it could even be detrimental since it's likely to drift out of sync and have conflicting or missing information.

We could expand the interface documentation though. `Accept` and `Dial` have basically the same semantics as e.g. the Go stdlib methods, so I'd think they're well-established? The current docs are:

```go
	// Accept waits for the next inbound connection on a listening endpoint.
	Accept(context.Context) (Connection, error)

	// Dial creates an outbound connection to an endpoint.
	Dial(context.Context, Endpoint) (Connection, error)
```

What do you feel is missing here?

Reply:
I feel like that is clear enough but requires me to jump to the interface. While I do agree with you, typically I still document the implementing methods because there might be additional context that would be beneficial for the reader.

If in this case the implementation requires no additional context that would be beneficial, feel free to resolve this thread 👍

Reply:
It's a convention that I was hoping to sneak in, but I don't think the rest of the code base follows it, so I'm happy to duplicate the docs if that's what we'd prefer.

Reply:
If there is no additional context you think is beneficial that is specific to `MConnTransport` then we're good 🎉